fix: Validate requestRedirect with backendrefs

slayer321 opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:

1. If this is your first time contributing to Gateway API, please read our
   developer guide (https://gateway-api.sigs.k8s.io/contributing/devguide/)
   and our community page (https://gateway-api.sigs.k8s.io/contributing/).
2. If this is your first time contributing to a Kubernetes project, please read
   our contributor guidelines:
   https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution
3. Please label this pull request according to what type of issue you are
   addressing, especially if this is a release targeted pull request. For
   reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
4. If you want *faster* PR reviews, read how:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
6. If this PR includes a new GEP please make sure you've followed the process
   outlined in our GEP overview, as this will help the community to ensure the
   best chance of positive outcomes for your proposal:
   https://gateway-api.sigs.k8s.io/geps/overview/#process
-->

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind gep
/kind test

Optionally add one or more of the following kinds if applicable:
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/area conformance
-->
/kind bug
/kind test
**What this PR does / why we need it**:
Validate and  Prevent Redirect Filters and BackendRefs in same HTTPRoute Rules.
**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #2148

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
Webhook validation now ensures that BackendRefs can not be specified in the same HTTPRoute rule as a Redirect filter
```

gyohuangxin commented:
/ok-to-test

gyohuangxin commented:
@slayer321 The job failed due to a lint error:
```
./hack/invalid-examples/v1beta1/httproute/invalid-request-redirect-with-backendref.yaml
  16:19     error    no new line character at the end of file  (new-line-at-end-of-file)
[0;31mTest FAILED: hack/../hack/verify-yamllint.sh 
make: *** [Makefile:114: verify] Error 1
```

gyohuangxin reviewed (approved):
LGTM, thanks!

arkodg commented on apis/v1alpha2/validation/httproute_test.go:
```diff
@@ -868,7 +868,7 @@ func TestValidateHTTPRouteTypeMatchesField(t *testing.T) {
 			},
 		},
-		errCount: 0,
+		errCount: 1,
```
why has this changed ?

slayer321 replied:
The same reason that I mentioned for `v1beta1` changes.

robscott replied:
@slayer321 do you happen to have a link handy to that comment or can you just copy it here? Not finding the context easily and also have the same question.

slayer321 replied:
I was taking about this comment https://github.com/kubernetes-sigs/gateway-api/pull/2161#discussion_r1251477315

arkodg commented on apis/v1beta1/validation/httproute_test.go:
```diff
@@ -1049,7 +1075,7 @@ func TestValidateHTTPRouteTypeMatchesField(t *testing.T) {
 			},
 		},
-		errCount: 0,
+		errCount: 1,
```
why has this changed ?

slayer321 replied:
So, as in this issue we need to validate that BackendRefs is not used when Redirect Filters  are used in httpRoute. 
In this [example](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/validation/httproute_test.go#L1041) as we can see that we are using Redirect Filters and then in the for loops we have [BackendRefs](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/validation/httproute_test.go#L1125) as well. 

After I added a check of it inside `validation/httproute.go`. The test started failing so I made the `errCount` as 1 .

arkodg replied:
ah thanks for the clarification, didnt go through at the entire file earlier 🙈 
would be good to add a positive test case, maybe as a new standalone function that ensures thats when the Redirect filter exists and no backendRefs are set, validation is successful

arkodg replied:
was referring to a standalone alone test `TestValidateRequestRedirectFiltersWithNoBackendRef`
anyways this is a good to have and is non blocking

arkodg commented on apis/v1beta1/validation/httproute.go:
```diff
@@ -77,6 +78,17 @@ func ValidateHTTPRouteSpec(spec *gatewayv1b1.HTTPRouteSpec, path *field.Path) fi
+	var errs field.ErrorList
+	for _, filter := range rule.Filters {
+		if filter.RequestRedirect != nil && len(rule.BackendRefs) > 0 {
+			errs = append(errs, field.Invalid(path.Child("filters"), filter, "RequestRedirect filter is not allowed with backendRefs"))
```
is `filter,` needed ?

slayer321 replied:
If you are asking if `filter` needed inside the `field.Invalid` function. So yeah I think we can change to `gatewayv1b1.HTTPRouteFilterRequestRedirect` .. as it will be more specific.
```
errs = append(errs, field.Invalid(path.Child("filters"), gatewayv1b1.HTTPRouteFilterRequestRedirect, "RequestRedirect filter is not allowed with backendRefs"))
```

arkodg replied:
thanks, yeah +1 to this change

arkodg reviewed (approved):
LGTM thanks !

slayer321 commented:
Hi, as api-verify CI is failing so I tried running the `golangci-lint run` command .. but it is just showing Killed 
```
sachinmaurya@sachin:~/go/src/github.com/gateway-api (test-request-route-w-backend-ref)$ golangci-lint run 
WARN [runner] The linter 'structcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter.  Replaced by unused. 
WARN [runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter.  Replaced by unused. 
WARN [runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter.  Replaced by unused. 
Killed
```
What can be the reason for that :thinking:

arkodg commented:
> golangci-lint run

@slayer321  can you rebase to `main` and try `./hack/verify-golint.sh` ?

slayer321 commented:
I tried it but still got below error 
```
sachinmaurya@sachin:~/go/src/github.com/gateway-api (test-request-route-w-backend-ref)$ ./hack/verify-golint.sh
level=warning msg="[runner] Can't run linter goanalysis_metalinter: musttag: running `go list all`: exit status 1"
level=error msg="Running error: 1 error occurred:\n\t* can't run linter goanalysis_metalinter: musttag: running `go list all`: exit status 1\n\n"
```
After looking for it , I found that there where many similar errors but nothing same as this one. most of them says that we need to run `go build`.
has anyone else faced the similar error while working on this project or is it some issue with my local setup.

gyohuangxin commented:
@slayer321 I don't have idea about it. Even if you can't run the `./hack/verify-golint.sh` in your environment, you can find the details from the CI job logs: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_gateway-api/2161/pull-gateway-api-verify/1676963399549849600

robscott commented:
/release-note "Webhook validation now ensures that BackendRefs can not be specified in the same HTTPRoute rule as a Redirect filter"

robscott commented:
/release-note-edit ```Webhook validation now ensures that BackendRefs can not be specified in the same HTTPRoute rule as a Redirect filter```

robscott commented:
Thanks @slayer321!

/lgtm
/approve
